File: src/world.py
"""Define World class."""

import importlib
import random
import functools

import entity
import config


class WorldInterface:
    pass


class World:
    def __init__(self, world_name, entity_names):
        # Define vars
        self.size = None
        self.entities = None
        self.energy = None
        self.height = None
        self.entity = None
        # Construct world interface
        self.interface = WorldInterface()
        self.interface.set_size = self.set_size
        self.interface.get_size = self.get_size
        self.interface.set_height = self.set_height
        self.interface.get_height = self.get_height
        self.interface.set_energy = self.set_energy
        self.interface.get_energy = self.get_energy
        # Load user defined functions
        tmp = importlib.import_module("worlds.{0}.main".format(world_name))
        self.user_init = tmp.init
        self.user_update = tmp.update
        # Initialize the world
        self.init()
        # Delete interfaces that should only be used in initialization
        del self.interface.set_size
        # Create entities
        # It will be appended to self.entities in self.create_entity, so don't worry
        self.entities = []
        for en in entity_names:
            self.create_entity(en)

        self.time = 0

    def init(self):
        """Do some necessary initialization, and then call user_init."""
        # No necessary initialization for now
        #
        # Initialize by user
        self.user_init(self.interface)

    def update(self):
        """Do some necessary update, and then call user_update."""
        # Necessary updates
        self.time += 1
        # Update by user
        self.user_update(self.interface)
        # Let every entity act
        for e in self.entities:
            e.act()

    def create_entity(self, entity_name):
        """Create an entity by name."""
        en = entity.Entity(entity_name)
        # Set initial position and energy
        # Don't wanna judge whether there is any position empty for now
        while True:
            x = random.randint(0, self.size[0]-1)
            y = random.randint(0, self.size[1]-1)
            if self.entity[x][y] is None:
                en.set_pos(x, y)
                self.entity[x][y] = True
                break
        en.set_energy(config.initial_energy())
        self.create_entity_interface(en)
        # Save current entity
        self.entities.append(en)

    def create_entity_interface(self, en):
        en.interface = entity.EntityInterface()

        # For convenience
        def pack(f):
            return functools.partial(f, en=en)
        # Actions
        # I don't think this is a beautiful implementation
        en.interface.move_up = pack(self.act_move_up)
        en.interface.move_down = pack(self.act_move_down)
        en.interface.move_left = pack(self.act_move_left)
        en.interface.move_right = pack(self.act_move_right)
        en.interface.eat = pack(self.act_eat)
        # Sensors
        en.interface.sense_height = pack(self.act_sense_height)
        en.interface.sense_energy = pack(self.act_sense_energy)
        en.interface.sense_world_size = pack(self.act_sense_world_size)
        en.interface.sense_pos = pack(self.act_sense_pos)

    def set_size(self, sx, sy):
        """Set world size to (sx, sy) and reset height and energy to 0 at each position."""
        self.size = (sx, sy)
        self.height = [[0]*sy]*sx
        self.energy = [[0]*sy]*sx
        self.entity = [[None]*sy]*sx

    def get_size(self):
        return self.size

    def set_height(self, x, y, h):
        self.height[x][y] = h

    def get_height(self, x, y):
        return self.height[x][y]

    def set_energy(self, x, y, e):
        self.energy[x][y] = e

    def get_energy(self, x, y):
        return self.energy[x][y]

    def act_eat(self, en):
        pass

    def act_move_up(self, en):
        pass

    def act_move_down(self, en):
        pass

    def act_move_left(self, en):
        pass

    def act_move_right(self, en):
        pass

    def act_sense_world_size(self, en):
        pass

    def act_sense_energy(self, en):
        pass

    def act_sense_height(self, en):
        pass

    def act_sense_pos(self, en):
        pass

File: src/world.cpp
#include <dlfcn.h>

#include "world.h"
#include "action.h"
#include "config.h"
#include "utils.h"

World::World(const char *name) {
    void *handle = dlopen(name, RTLD_LAZY);
    if (!handle) {
        crash("failed loading shared library");
    }
    init = (void(*)(World&))dlsym(handle, "init");
    update = (void(*)(World&))dlsym(handle, "update");
    if (!init || !update) {
        crash("invalid shared library for world");
    }
    sizeX = sizeY = -1;
    init(*this);
    if (sizeX < 0 || sizeY < 0) {
        crash("invalid shared library for world");
    }
}

World::~World() {
}

void World::Update() {
    update(*this);
    for (auto &e: entities) {
        if (e.energy <= 0) continue;
        e.energy -= config::update_cost;
        Action action = e.Act();
        TakeAction(e, action);
    }
}

void World::CreateEntity(const char *name) {
    Entity entity(name);
    int x, y;
    do {
        x = rand()%sizeX;
        y = rand()%sizeY;
    } while(grids[x][y].occupied);
    grids[x][y].occupied = true;
    entity.SetPosition(x, y);
   
    entity.energy = config::initial_energy;

    // Set interface functions
    entity.SenseEnergy = [this](int x, int y){ return this->GetEnergy(x, y); };
    entity.SenseHeight = [this](int x, int y){ return this->GetHeight(x, y); };
    entity.SenseWorldSize = [this](int &x, int &y){ this->GetSize(x, y);};
    
    entity.init(entity);

    entities.push_back(entity);
}

void World::TakeAction(Entity &entity, Action action) {
    switch(action.type) {
        case Move:
            switch(action.param.direction) {
                case Up:
                    TakeActionMove(entity, 0, -1);
                    break;
                case Down:
                    TakeActionMove(entity, 0, +1);
                    break;
                case Left:
                    TakeActionMove(entity, -1, 0);
                    break;
                case Right:
                    TakeActionMove(entity, 1, 0);
                    break;
            }
            break;
        case Eat:
            TakeActionEat(entity);
            break;
    }
}

void World::TakeActionMove(Entity &entity, int dx, int dy) {
    int tx = entity.posiX + dx;
    int ty = entity.posiY + dy;
    if (0 <= tx && tx < sizeX
            && 0 <= ty && ty < sizeY
            && !grids[tx][ty].occupied) {
        grids[tx][ty].occupied = true;
        grids[entity.posiX][entity.posiY].occupied = false;

        entity.energy -= config::move_cost(grids[tx][ty].height - grids[entity.posiX][entity.posiY].height);
        entity.SetPosition(tx, ty);

        entity.SensePosition(tx, ty);
    }
}

void World::TakeActionEat(Entity &entity) {
    entity.energy += grids[entity.posiX][entity.posiY].energy;
    grids[entity.posiX][entity.posiY].energy = 0;
}

void World::SetSize(int sizeX, int sizeY) {
    this->sizeX = sizeX;
    this->sizeY = sizeY;
    grids = vector< vector< Grid > >(sizeX, vector<Grid>(sizeY));
}

void World::SetHeight(int x, int y, int h) {
    grids[x][y].height = h;
}

void World::SetEnergy(int x, int y, int e) {
    grids[x][y].energy = e;
}

void World::GetSize(int &sizeX, int &sizeY) const {
    sizeX = this->sizeX;
    sizeY = this->sizeY;
}

int World::GetHeight(int x, int y) const {
    return grids[x][y].height;
}

int World::GetEnergy(int x, int y) const {
    return grids[x][y].energy;
}

const vector< Entity > & World::GetEntities() const{
    return entities;
}

bool World::IsOccupied(int x, int y) const {
    return grids[x][y].occupied;
}
